File: 0200-number-of-islands/0200-number-of-islands.py
class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:
        if not grid:
            return 0
        
        num_rows, num_cols = len(grid), len(grid[0])
        visited = set()

        def get_neighbors(coord):
            row, col = coord
            delta_row = [-1, 0, 1, 0]
            delta_col = [0, 1, 0, -1]
            res = []
            for i in range(len(delta_row)):
                neighbor_row = row + delta_row[i]
                neighbor_col = col + delta_col[i]
                if 0 <= neighbor_row < num_rows and 0 <= neighbor_col < num_cols:
                    res.append((neighbor_row, neighbor_col))
            return res

        def bfs(starting_node):
            queue = deque([starting_node])
            visited.add(starting_node)
            while queue:
                node = queue.popleft()
                for neighbor in get_neighbors(node):
                    r, c = neighbor
                    if grid[r][c] == '1' and neighbor not in visited:
                        queue.append(neighbor)
                        visited.add(neighbor)

        island_count = 0

        # Iterate through the grid
        for r in range(num_rows):
            for c in range(num_cols):
                if grid[r][c] == '1' and (r, c) not in visited:
                    bfs((r, c))  # Perform BFS to mark the entire island
                    island_count += 1
        
        return island_count

File: 0200-number-of-islands/0200-number-of-islands.java
class Solution {
    public int numIslands(char[][] grid) {
        int rows = grid.length;
        int columns = grid[0].length;
        int count = 0;
        boolean[][] visited = new boolean[rows][columns];

        for (int i = 0; i < rows; i++) {
            for (int j = 0; j < columns; j++) {
                if (grid[i][j] == '1' && !visited[i][j]) {
                    bfs(grid, i, j, visited, rows, columns);
                    count++;
                }
            }
        }
        return count;
    }

    public void bfs(char[][] grid, int i, int j, boolean[][] visited, int rows, int cols) {
        Queue<int[]> queue = new LinkedList<>();
        queue.add(new int[]{i, j});          // Bug 2 fix: i,j not r,c
        visited[i][j] = true;

        int[][] directions = {{-1,0},{0,1},{1,0},{0,-1}};  // Bug 3 fix: semicolon

        while (!queue.isEmpty()) {
            int[] cell = queue.poll();
            int row = cell[0];
            int col = cell[1];

            for (int[] dir : directions) {
                int nr = row + dir[0];       // Bug 4 fix: dir[0] not dr
                int nc = col + dir[1];       // Bug 4 fix: dir[1] not dc

                if (nr >= 0 && nr < rows && nc >= 0 && nc < cols
                        && !visited[nr][nc] && grid[nr][nc] == '1') {
                    queue.add(new int[]{nr, nc});   // Bug 5 fix: new int[]
                    visited[nr][nc] = true;
                }
            }
        }                                    // Bug 6 fix: closing braces
    }
}